Accept transforms list instead of matrix for transform view parameter.

kmagiera opened the pull request:
In #7916 I moved transform matrix decomposition logic from JS to java. The next step is to accept list of transforms instead oftransform matrix as a transform @ReactProp. This way there is no extra processing required on JS side for the transform param (at least for android now) and this on the other hand allow us to execute transform updates (through offloaded animation) solely on the UI thread.

After this change there is a whole bunch of stuff from `Libraries/Utilities/MatrixMath.js` that can be deleted (methods like: determinant, inverse, transpose). Although @astreet mentioned under one of my previous commits that the code is still being referenced internally at fb, so I decided not to delete it here.

**Test plan (required)**
Run UIExplorer Transform example before and after - compare the results
Run android unit test: com.facebook.react.uimanager.MatrixMathHelperTest

facebook-github-bot-7 commented:
By analyzing the blame information on this pull request, we identified @sebmarkbage and @asmockler to be potential reviewers.

facebook-github-bot-7 commented:
Thank you for your pull request.  As you may know, we require contributors to sign our Contributor License Agreement, and we don't seem to have you on file and listed as active anymore.  In order for us to review and merge your code, please email cla@fb.com with your details so we can update your status.

janicduplessis commented on Libraries/StyleSheet/processTransform.js:
```diff
@@ -26,7 +26,12 @@ var stringifySafe = require('stringifySafe');
+  // Android implementation of transform property accepts the list of transform
+  // properties as opposed to a transform Matrix. This is necessary to control
+  // transform property updates completely on the native thread.
+  if (Platform.OS === 'android') {
```
We are doing a bunch of dev validations that won't be executed anymore and that are not in the java code, maybe move the validations in another loop before returning on android.

brentvatne replied:
@janicduplessis - is this blocking landing this PR?

janicduplessis replied:
Yea I think we should fix this before shipping this to avoid regressing on dev experience.

brentvatne commented:
@kmagiera @astreet - anything blocking us from landing this?

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/TransformHelper.java:
```diff
@@ -0,0 +1,91 @@
+import com.facebook.react.bridge.ReadableMap;
+import com.facebook.react.bridge.ReadableType;
+
+/**
```
Put comments

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/TransformHelper.java:
```diff
@@ -0,0 +1,91 @@
+ */
+public class TransformHelper {
+
+  private static final double[] HELPER_MATRIX = new double[16];
```
Can we actually make this a ThreadLocal? I'm concerned that we will want to do this math both on the shadow and main threads at some point and this isn't currently threadsafe.

astreet commented:
lgtm besides inlines

kmagiera commented:
@astreet all comments have been addressed:
 - extracted transform validation in JS and now is executed in DEV mode on android and iOS
 - use thread local for the helper matrix in `TransformHelper.java`
 - added comment in `TransformHelper.java`

facebook-github-bot-3 commented:
@kmagiera updated the pull request.

asmockler commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/MatrixMathHelper.java:
```diff
@@ -350,4 +381,79 @@ public static void quaternionToDegreesXYZ(double[] q, double[] result) {
+  }
+
+  public static void applyPerspective(double[] m, double perspective) {
+    m[11] = -1 / perspective;
```
This could cause a divide by 0 error

kmagiera replied:
yes, in which case it would trigger an appropriate exception, I don't see a point of someone setting perspective to 0 so I don't think there is a need to handle that case gracefully. This value is directly extracted from the transform matrix which validated in DEV mode (see [processTransform.js#168](https://github.com/facebook/react-native/blob/a59afb98d5aba8c4826eda349a4432457aab1d4d/Libraries/StyleSheet/processTransform.js#L168))

asmockler replied:
Sounds good. Sometimes the value will unintentionally pass over 0 during an animation, but I agree this is probably not the time or the place to handle for that. Just thought I would mention it. Thanks :)

facebook-github-bot-2 commented:
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

facebook-github-bot-3 commented:
@kmagiera updated the pull request.

facebook-github-bot-3 commented:
@kmagiera updated the pull request.

astreet commented:
@facebook-github-bot shipit

facebook-github-bot-8 commented:
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3676017) to review internal test results.
